### PR TITLE
refactor shopping list generation

### PR DESCRIPTION
**Pull request by xbohov01**

### CODE REVIEW

File: src/components/loadout/LoadoutExporter.tsx [hunk 1]
```diff
@@ -4,11 +4,11 @@ import { Button, VStack } from "@chakra-ui/react";
 
 import LoadoutExportText from "./LoadoutExportText";
 
-import type { ListKey, LocatedItem } from "~type/search";
+import type { LocatedItem, ShoppingListLocationKey } from "~type/search";
 import type { KeyValue } from "~type/select";
 
 type LoadoutExporterProps = {
-  gear: KeyValue<ListKey, LocatedItem[]>[];
+  gear: KeyValue<ShoppingListLocationKey, LocatedItem[]>[];
```
Comment (thatleda) [inline src/components/loadout/LoadoutExporter.tsx:11]: I think we could benefit from just using Record<string, LocatedItem[]>. If the salelocations are identified by a compound string, it's much faster to run `uniq` on.

File: src/components/shoppingList/ShoppingList.tsx [hunk 2]
```diff
@@ -8,25 +8,29 @@ import shoppingListGenerator from "../shoppingListGenerator";
...
 
   useEffect(() => {
     try {
       shoppingListGenerator.GetShoppingList(props.gear).then((res) => {
-        setPrice(res[0]);
-        setList(res[1]);
-        props.listUpstream(res[1]);
+        setPrice(res.totalPrice);
+        setList(res.locatedItems);
+        props.listUpstream(res.locatedItems);
```
Comment (thatleda) [inline src/components/shoppingList/ShoppingList.tsx:33]: Sounds like `onGearChange` or `onGearUpdate` to me. Don't enjoy the word `upsteam` for this.

File: src/components/shoppingListGenerator.ts [hunk 3]
```diff
@@ -1,107 +1,102 @@
...
     );
     const results = await Promise.all(promises);
 
-    results.forEach((result) => {
-      if (!result.success && result.message !== "Not sold") {
-        throw new Error(`SaleLocationsFetchException: ${result.message}`);
-      }
-
-      price += result.data.length > 0 ? result.data[0].price : 0;
+    const resultData = results.flatMap((r) => r.data);
```
Comment (thatleda) [inline src/components/shoppingListGenerator.ts:21]: At some point I will turn `noImplicitAny` on to punish you for this. Don't shorten iterators to one letter either, we're not saving on the energy bill.

File: package-lock.json [hunk 4]
```diff
@@ -1,14 +1,10673 @@
 {
```
Comment (thatleda) [inline package-lock.json:1]: Why did this file update?

**Review by thatleda — COMMENTED** [review]
Good job simplifying and specifying! :)

File: src/components/loadout/LoadoutExporter.tsx (hunk 1 above)
Comment (xbohov01) [reply src/components/loadout/LoadoutExporter.tsx:11]: Record only takes string, this case needs the compound object key

File: src/components/shoppingListGenerator.ts (hunk 3 above)
Comment (xbohov01) [reply src/components/shoppingListGenerator.ts:21]: Is not an implicit any, results are typed

**Pull request merged**